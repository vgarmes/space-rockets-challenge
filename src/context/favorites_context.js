import React, { useEffect, useContext, useReducer } from 'react';
import reducer from '../reducers/favorites_reducer';
import {
  ADD_FAVORITE_LAUNCH,
  REMOVE_FAVORITE_LAUNCH,
  ADD_FAVORITE_LAUNCHPAD,
  REMOVE_FAVORITE_LAUNCHPAD,
  TOGGLE_EXPANDED_MENU_ITEM,
} from '../constants/actions';

const getLocalStorage = () => {
  let favorites = localStorage.getItem('favorites');
  if (favorites) {
    return JSON.parse(localStorage.getItem('favorites'));
  } else {
    return { launches: [], launch_pads: [] };
  }
};

const initialState = {
  favorites: getLocalStorage(),
  expandedMenuItems: [false, false],
};

const FavoritesContext = React.createContext();

export const FavoritesProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const addFavoriteLaunch = (id, name) => {
    dispatch({
      type: ADD_FAVORITE_LAUNCH,
      payload: { id, name },
    });
  };

  const removeFavoriteLaunch = (id) => {
    dispatch({
      type: REMOVE_FAVORITE_LAUNCH,
      payload: { id },
    });
  };

  const addFavoriteLaunchPad = (id, name) => {
    dispatch({
      type: ADD_FAVORITE_LAUNCHPAD,
      payload: { id, name },
    });
  };

  const removeFavoriteLaunchPad = (id) => {
    dispatch({
      type: REMOVE_FAVORITE_LAUNCHPAD,
      payload: { id },
    });
  };

  const toggleExpandedMenuItem = (index) => {
    console.log(index);
    dispatch({
      type: TOGGLE_EXPANDED_MENU_ITEM,
      payload: index,
    });
  };

  useEffect(() => {
    localStorage.setItem('favorites', JSON.stringify(state.favorites));
  }, [state.favorites]);

  return (
    <FavoritesContext.Provider
      value={{
        ...state,
        addFavoriteLaunch,
        removeFavoriteLaunch,
        addFavoriteLaunchPad,
        removeFavoriteLaunchPad,
        toggleExpandedMenuItem,
      }}
    >
      {children}
    </FavoritesContext.Provider>
  );
};

export const useFavoritesContext = () => {
  return useContext(FavoritesContext);
};
