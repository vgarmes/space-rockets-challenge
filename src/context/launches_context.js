import React, { useContext, useReducer } from 'react';
import reducer from '../reducers/launches_reducer';
import {
  UPDATE_SORT,
  SET_GRIDVIEW,
  SET_LISTVIEW,
  UPDATE_FILTERS,
} from '../constants/actions';

const initialState = {
  sort: 'launch_date_utc',
  order: 'desc',
  grid_view: true,
  filters: {
    launch_success: ['successful', 'failed'],
    site_id: 'all',
  },
};

const LaunchesContext = React.createContext();

export const LaunchesProvider = ({ children }) => {
  const [state, dispatch] = useReducer(reducer, initialState);

  const updateSort = (value) => {
    dispatch({ type: UPDATE_SORT, payload: value });
  };

  const setGridView = () => {
    dispatch({ type: SET_GRIDVIEW });
  };

  const setListView = () => {
    dispatch({ type: SET_LISTVIEW });
  };

  const updateFilters = (name, value) => {
    dispatch({ type: UPDATE_FILTERS, payload: { name, value } });
  };

  return (
    <LaunchesContext.Provider
      value={{ ...state, updateSort, setGridView, setListView, updateFilters }}
    >
      {children}
    </LaunchesContext.Provider>
  );
};

export const useLaunchesContext = () => {
  return useContext(LaunchesContext);
};
