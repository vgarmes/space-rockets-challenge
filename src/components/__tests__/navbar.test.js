import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import { FavoritesProvider } from '../../context/favorites_context';
import { ThemeProvider, CSSReset } from '@chakra-ui/core';
import { Navbar } from '../';

// baseElement has to be snapshotted because the drawer is a modal that renders outside Navbar fragment
test('drawer is closed by default', () => {
  const { baseElement } = render(
    <Router>
      <FavoritesProvider>
        <ThemeProvider>
          <CSSReset />
          <Navbar />
        </ThemeProvider>
      </FavoritesProvider>
    </Router>
  );

  expect(baseElement).toMatchSnapshot();
});

test('drawer slides in on menu click', () => {
  const { getByRole, baseElement } = render(
    <Router>
      <FavoritesProvider>
        <ThemeProvider>
          <CSSReset />
          <Navbar />
        </ThemeProvider>
      </FavoritesProvider>
    </Router>
  );

  fireEvent.click(getByRole('button', { name: /menu/i }));

  expect(baseElement).toMatchSnapshot();
});
