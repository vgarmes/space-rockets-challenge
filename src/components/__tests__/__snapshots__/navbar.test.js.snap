// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`drawer is closed by default 1`] = `
<body>
  <div
    class="Toaster"
    id="react-toast"
  >
    <span
      class="Toaster__manager-top"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; top: 0px; right: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-top-left"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; top: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-top-right"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; top: 0px; right: 0px;"
    />
    <span
      class="Toaster__manager-bottom-left"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-bottom"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; bottom: 0px; right: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-bottom-right"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; right: 0px;"
    />
  </div>
  <div>
    <nav
      class="css-vn1rpq"
    >
      <p
        class="css-vjjv55"
      >
        ¡SPACE·R0CKETS!
      </p>
      <button
        class="css-1wyike"
        type="button"
      >
        Menu
      </button>
    </nav>
  </div>
</body>
`;

exports[`drawer slides in on menu click 1`] = `
<body>
  <div
    aria-hidden="true"
    class="Toaster"
    data-aria-hidden="true"
    id="react-toast"
  >
    <span
      class="Toaster__manager-top"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; top: 0px; right: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-top-left"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; top: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-top-right"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; top: 0px; right: 0px;"
    />
    <span
      class="Toaster__manager-bottom-left"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-bottom"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; margin: 0px auto; text-align: center; bottom: 0px; right: 0px; left: 0px;"
    />
    <span
      class="Toaster__manager-bottom-right"
      style="max-width: 560px; position: fixed; z-index: 5500; pointer-events: none; bottom: 0px; right: 0px;"
    />
  </div>
  <div
    aria-hidden="true"
    data-aria-hidden="true"
  >
    <nav
      class="css-vn1rpq"
    >
      <p
        class="css-vjjv55"
      >
        ¡SPACE·R0CKETS!
      </p>
      <button
        class="css-1wyike"
        type="button"
      >
        Menu
      </button>
    </nav>
  </div>
  <div
    id="chakra-portal-1"
  >
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="1"
    />
    <div
      data-focus-lock-disabled="false"
    >
      <div
        class="css-1i9whq8"
      />
      <div
        class="css-556rku"
      >
        <section
          aria-describedby="drawer-1-body"
          aria-labelledby="drawer-1-header"
          aria-modal="true"
          class="css-105zo18"
          id="drawer-1"
          role="dialog"
          style="will-change: opacity, transform; transform: translateX(100%); right: 0px; top: 0px; height: 100vh;"
          tabindex="-1"
        >
          <button
            aria-label="Close"
            class="css-1dedo85"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="css-1idynds"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M.439,21.44a1.5,1.5,0,0,0,2.122,2.121L11.823,14.3a.25.25,0,0,1,.354,0l9.262,9.263a1.5,1.5,0,1,0,2.122-2.121L14.3,12.177a.25.25,0,0,1,0-.354l9.263-9.262A1.5,1.5,0,0,0,21.439.44L12.177,9.7a.25.25,0,0,1-.354,0L2.561.44A1.5,1.5,0,0,0,.439,2.561L9.7,11.823a.25.25,0,0,1,0,.354Z"
                fill="currentColor"
              />
            </svg>
          </button>
          <header
            class="css-w7idnj"
            id="drawer-1-header"
          >
            Menu
          </header>
          <div
            class="css-5m10fw"
            id="drawer-1-body"
          >
            <div
              class="css-0"
              data-accordion=""
            >
              <div
                class="css-1g0j2h0"
                data-accordion-item=""
              >
                <button
                  aria-controls="accordion-panel-3"
                  aria-expanded="false"
                  class="css-mny6m6"
                  id="accordion-header-3"
                  type="button"
                >
                  <div
                    class="css-10d20og"
                  >
                    Launches
                  </div>
                  <svg
                    aria-hidden="true"
                    class="css-1duxe6l"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
                <div
                  aria-hidden="true"
                  class="rah-static rah-static--height-zero css-vppuuz"
                  style="height: 0px; overflow: hidden;"
                >
                  <div
                    style="transition: opacity 250ms ease 0ms; opacity: 0; display: none;"
                  >
                    <div
                      aria-hidden="true"
                      aria-labelledby="accordion-header-3"
                      class="css-1oxb0in"
                      data-accordion-panel=""
                      id="accordion-panel-3"
                      role="region"
                    >
                      <a
                        class="css-1hvndql"
                        href="/launches"
                      >
                        Add favorites...
                      </a>
                    </div>
                  </div>
                </div>
              </div>
              <div
                class="css-1g0j2h0"
                data-accordion-item=""
              >
                <button
                  aria-controls="accordion-panel-4"
                  aria-expanded="false"
                  class="css-mny6m6"
                  id="accordion-header-4"
                  type="button"
                >
                  <div
                    class="css-10d20og"
                  >
                    Launch pads
                  </div>
                  <svg
                    aria-hidden="true"
                    class="css-1duxe6l"
                    focusable="false"
                    role="presentation"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
                      fill="currentColor"
                    />
                  </svg>
                </button>
                <div
                  aria-hidden="true"
                  class="rah-static rah-static--height-zero css-vppuuz"
                  style="height: 0px; overflow: hidden;"
                >
                  <div
                    style="transition: opacity 250ms ease 0ms; opacity: 0; display: none;"
                  >
                    <div
                      aria-hidden="true"
                      aria-labelledby="accordion-header-4"
                      class="css-1oxb0in"
                      data-accordion-panel=""
                      id="accordion-panel-4"
                      role="region"
                    >
                      <a
                        class="css-1hvndql"
                        href="/launch-pads"
                      >
                        Add favorites...
                      </a>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
          <footer
            class="css-81g6me"
          >
            Drawer footer
          </footer>
        </section>
      </div>
    </div>
    <div
      data-focus-guard="true"
      style="width: 1px; height: 0px; padding: 0px; overflow: hidden; position: fixed; top: 1px; left: 1px;"
      tabindex="0"
    />
  </div>
</body>
`;
